subprojects/core/src/main/java/org/gradle/internal/classpath/CallInterceptionClosureInstrumentingClassVisitor.java

```diff
@@ -112,25 +113,50 @@ public MethodVisitorScopeImpl(MethodVisitor methodVisitor) {
                     public void visitCode() {
                         /*
                          * enterInstrumentedClosure(this);
-                         * doCall$original(<args>);
-                         * leaveInstrumentedClosure(this);
+                         * try {
+                         *     return doCall$original(<args>);
+                         * } finally { // but with inlining the handler
+                         *     leaveInstrumentedClosure(this);
+                         * }
                          */
                         _ALOAD(0);
                         String enterLeaveDescriptor = Type.getMethodDescriptor(Type.VOID_TYPE, getType(InstrumentableClosure.class));
                         _INVOKESTATIC(Type.getType(InstrumentedClosuresHelper.class).getInternalName(), "enterInstrumentedClosure", enterLeaveDescriptor, false);
 
+                        Label tryBlockStart = new Label();
+                        Label tryBlockEnd = new Label();
+                        Label catchBlockStart = new Label();
+
+                        mv.visitTryCatchBlock(tryBlockStart, tryBlockEnd, catchBlockStart, "java/lang/Throwable");
+
+                        mv.visitLabel(tryBlockStart);
                         _ALOAD(0); // receiver reference to this
                         // Invoke the original method:
                         Type[] argumentTypes = Type.getArgumentTypes(methodData.descriptor);
                         for (int argIndex = 1; argIndex <= argumentTypes.length; ++argIndex) {
                             visitVarInsn(argumentTypes[argIndex - 1].getOpcode(Opcodes.ILOAD), argIndex);
                         }
                         _INVOKESPECIAL(clazz.className, methodNameToVisit, methodData.descriptor, false);
+                        mv.visitLabel(tryBlockEnd);
 
+                        // finally:
                         _ALOAD(0);
                         _INVOKESTATIC(Type.getType(InstrumentedClosuresHelper.class).getInternalName(), "leaveInstrumentedClosure", enterLeaveDescriptor, false);
-
+                        // and return:
                         visitInsn(Type.getReturnType(methodData.descriptor).getOpcode(IRETURN));
+
+                        // start exception handler:
+                        mv.visitLabel(catchBlockStart);
+                        Object[] locals = new Object[]{clazz.className.replaceAll("\\.", "/")};
+                        // Must use an F_NEW frame, as we may encounter class versions <= V1_5, see ASM MethodWriter
+                        mv.visitFrame(Opcodes.F_NEW, 1, locals, 1, new Object[]{"java/lang/Throwable"});
+
+                        // finally:
```

Comment:
🤔  It is a bit confusing that there are two `// finally:` comments in the method. Is this intentional?

Reply:
Thanks, I've added a little more detail to the comments.